bip-0008.mediawiki

```diff
@@ -169,6 +164,29 @@ And ACTIVE and FAILED are terminal states, which a deployment stays in once they
         }
     }
 
+Note that blocks received while in MUST_SIGNAL must be checked to ensure that the threshold will be reached by the end of the period. For example:
+
+    if (GetStateForBlock(block) == MUST_SIGNAL) {
+        if (block.nVersion & 0xE0000000 != 0x2000000 || block.nVersion >> bit) & 1 != 1) {
+            nonsignal = 1;
+            count = block.nHeight % 2016;
+            walk = block;
+            while (count > 0 && nonsignal + count + threshold > 2016) {
+                --count;
+                walk = walk.parent;
+                if (walk.nVersion & 0xE0000000 != 0x2000000 || walk.nVersion >> bit) & 1 != 1) {
+                    ++nonsignal;
+                    if (nonsignal + threshold > 2016) break;
+                }
+            }
+            if (nonsignal + threshold > 2016) {
+                return state.Invalid(BlockValidationResult::RECENT_CONSENSUS_CHANGE, "blk-bip8-must-signal");
+            }
+        }
+    }
+
+Implementations should be careful not to ban peers that send blocks that are invalid due to not signalling (or blocks that build on those blocks), as that would allow an incompatible chain that is only briefly longer than the compliant chain to cause a split of the p2p network. If that occurred, nodes that are not enforcing lockinontimeout may not see new blocks in the compliant chain, and thus not reorg to it at the point when it has more work, and would thus not be following the valid chain with the most work.
```

Comment:
But even more importantly, nodes enforcing `MUST_SIGNAL` need to ensure they remain attached to other such nodes.

Reply:
I think the details of that probably need to be left to the bips that want to deploy via bip-8; they could ensure they remain connected to other nodes by making sure lockinontimeout only happens when everybody is happy with it, in which case no special code is needed; or they could allocate a temporary service flag bit for it and preferentially connect to other nodes that way for some period.

Still worth mentioning, but nothing's coming to mind as to what to say exactly.

Reply:
Added a note about this in a separate commit